state/protocol/protocol_state/kvstore/upgrade_statemachine.go

```diff
@@ -0,0 +1,152 @@
+	// We would like to guarantee that all nodes switch the protocol version at exactly the same block.
+	// To guarantee that all nodes switch the protocol version at exactly the same block, we require that the
+	// activation candidateView is higher than the sealing candidateView + Δ when accepting the event. Δ represents the finalization lag
+	// to give time for replicas to finalize the block containing the seal for the version upgrade event.
+	// When replica reaches activation candidateView and the latest finalized protocol state knows about the version upgrade,
+	// then it's safe to switch the protocol version.
+	if m.candidateView+m.params.EpochCommitSafetyThreshold() >= versionUpgrade.ActiveView {
+		return protocol.NewInvalidServiceEventErrorf("invalid protocol state version upgrade candidateView %d -> %d: %w",
+			m.candidateView+m.params.EpochCommitSafetyThreshold(), versionUpgrade.ActiveView, ErrInvalidActivationView)
+	}
+
+	if m.parentState.GetProtocolStateVersion() >= versionUpgrade.NewProtocolStateVersion {
+		return protocol.NewInvalidServiceEventErrorf("invalid protocol state version upgrade %d -> %d: %w",
+			m.parentState.GetProtocolStateVersion(), versionUpgrade.NewProtocolStateVersion, ErrInvalidUpgradeVersion)
+	}
+
+	// checkPendingUpgrade checks if there is a pending upgrade in the state and validates if we can set the new upgrade.
+	// We allow setting version upgrade if all the conditions are met:
+	// (i) the activation candidateView is higher than the current candidateView + Δ.
+	// (ii) if there is a pending upgrade, the new version should be the same as the pending upgrade.
+	// Condition (ii) is checked in this function.
+	checkPendingUpgrade := func(store protocol_state.KVStoreReader) error {
+		if pendingUpgrade := store.GetVersionUpgrade(); pendingUpgrade != nil {
+			if pendingUpgrade.ActivationView < m.candidateView {
+				// pending upgrade has been activated, we can ignore it.
+				return nil
+			}
+
+			// we allow updating pending upgrade iff the new version is the same as the pending upgrade
+			// the activation candidateView may differ, but it has to meet the same threshold.
+			if pendingUpgrade.Data != versionUpgrade.NewProtocolStateVersion {
+				return protocol.NewInvalidServiceEventErrorf("requested to upgrade to %d but pending upgrade with version already stored %d: %w",
+					pendingUpgrade.Data, versionUpgrade.NewProtocolStateVersion, ErrInvalidUpgradeVersion)
+			}
+		}
+		return nil
+	}
+
+	// check in case there is a pending upgrade in parent state.
+	err := checkPendingUpgrade(m.parentState)
+	if err != nil {
+		return fmt.Errorf("version upgrade invalid with respect to the parent state: %w", err)
+	}
+	// check in case there are multiple upgrades in the same block.
+	err = checkPendingUpgrade(m.mutator)
+	if err != nil {
+		return fmt.Errorf("version upgrade invalid with respect to the current state: %w", err)
+	}
```

Comment:
quite understand why we have to check the parent state. Aren't these all transitive checks? In other words, if I already have checked the compatibility of Activator1 ->  Activator2 -> ... -> ActivatorK and now I am finding that it is legal to modify ActivatorK -> ActivatorL. Then by transitivity, wouldn't it hold that I could directly go from Activator1 -> ActivatorL?

In other words, if we encounter multiple service events, we would write the newest one into `m.mutator` and if we haven't encountered any Service events, `m.mutator` contains the latest activator from the parent. In all cases, we only need to compare to `m.mutator` I think. 
 
```suggestion
	// There can be multiple `versionUpgrade` Service Events sealed in one block. In case we have _not_
	// encountered any, `m.mutator` contains the latest `versionUpgrade` as of the parent block, because
	// we cloned it from the parent state. If we encountered some version upgrades, we already enforced that
	// they are all upgrades to the same version. So we only need to check that the next `versionUpgrade`
	// also has the same version.
	err := checkPendingUpgrade(m.mutator)
	if err != nil {
		return fmt.Errorf("version upgrade invalid with respect to the current state: %w", err)
	}
```

Reply:
Yeah, if we consider the fact that replicated state is a copy of parent state then sure, we need to check only the current state. I haven't thought about this, applying change.